### PR TITLE
Add check for arguments

### PR DESCRIPTION
**Pull request by genestack-solomatin**

### CODE REVIEW

File: genestack_client/data_importer.py [hunk 1]
```diff
@@ -122,6 +122,19 @@ def __invoke_loader(self, parent, importer_type, metainfo):
         self.__process_links(metainfo)
         return self.importer.invoke('createFile', parent, importer_type, metainfo)['accession']
 
+    def __add_to_metainfo(self, metainfo, key, value, setter, required=False):
+        if value is None:
+            if required:
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: `if required and metainfo.get(key) is None:`?

File: genestack_client/data_importer.py [hunk 2]
```diff
@@ -122,6 +122,19 @@ def __invoke_loader(self, parent, importer_type, metainfo):
         self.__process_links(metainfo)
         return self.importer.invoke('createFile', parent, importer_type, metainfo)['accession']
 
+    def __add_to_metainfo(self, metainfo, key, value, setter, required=False):
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: I've written a couple of lines before realizing that I was wrong about some details behind the function =)
So, there's a need for doctsring!

File: genestack_client/data_importer.py [hunk 3]
```diff
@@ -122,6 +122,19 @@ def __invoke_loader(self, parent, importer_type, metainfo):
...
+        if value is None:
+            if required:
+                if metainfo.get(key) is None:
+                    raise GenestackException('Missed required key "%s", '
+                                             'it should be specified as function argument '
+                                             'or in metainfo' % key)
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: `'or in metainfo object'`

File: genestack_client/data_importer.py [hunk 4]
```diff
@@ -122,6 +122,19 @@ def __invoke_loader(self, parent, importer_type, metainfo):
...
+        else:
+            if metainfo.get(key) is not None:
+                raise GenestackException('Key "%s", is passed both as function argument '
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: Bad comma (`,`), remove it.

**Comment by klas-genestack** [comment]
The idea seems solid, but I'd rewrite the functions thus:
```py
    def __add_to_metainfo(self, metainfo, key, value, setter, required=False):
        mi_key = metainfo.get(key)
        if required and value is None and mi_key is None:
            raise GenestackException('Missing required key "%s", '
                                     'it should be passed as function argument '
                                     'or in metainfo object' % key)
        if value is not None and mi_key is not None:
            raise GenestackException('Key "%s" is passed both as function argument '
                                     'and inside metainfo object' % key)
        value = value if value is not None else mi_key
        setter(key, value)
```

File: genestack_client/data_importer.py [hunk 5]
```diff
@@ -124,6 +124,35 @@ def __invoke_loader(self, parent, importer_type, metainfo):
         self.__process_links(metainfo)
         return self.importer.invoke('createFile', parent, importer_type, metainfo)['accession']
 
+    def __add_to_metainfo(self, metainfo, key, value, setter, required=False):
+        """
+        Add value to metainfo, check if value is not already present,
+         checks required fields.
```
Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: remove space

File: genestack_client/data_importer.py (hunk 2 above)
Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: done

File: genestack_client/data_importer.py [hunk 6]
```diff
@@ -182,9 +214,11 @@ def create_vcf(self, parent=None, name=None, reference_genome=None, url=None, me
         :rtype: str
         """
         metainfo = metainfo or Metainfo()
-        name and metainfo.add_string(Metainfo.NAME, name)
+        self.__add_to_metainfo(metainfo, Metainfo.NAME, name,
+                               metainfo.add_string, required=True)
         reference_genome and metainfo.add_file_reference(BioMetaKeys.REFERENCE_GENOME, reference_genome)
```
Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: missed

**Comment by genestack-solomatin** [comment]
> The idea seems solid, but I'd rewrite the functions thus:

No, we should add value only if it is missed in metainfo

File: genestack_client/data_importer.py [hunk 7]
```diff
@@ -336,12 +375,14 @@ def create_sequencing_assay(self, parent, name=None, urls=None,
...
-        method and metainfo.add_string(BioMetaKeys.METHOD, method)
+        self.__add_to_metainfo(metainfo, Metainfo.NAME, name, StringValue, required=True)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.ORGANISM, organism, StringValue)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.METHOD, method, StringValue)
         if urls:
+            urls = (url for url in urls)
+            self.__add_to_metainfo(metainfo, BioMetaKeys.READS_LINK, urls.next(), ExternalLink)
             for url in urls:
-                metainfo.add_external_link(BioMetaKeys.READS_LINK, url)
+                metainfo.add_value(BioMetaKeys.READS_LINK, ExternalLink(url))
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: Do we intend to deprecate `add_internal_link()`?

File: genestack_client/data_importer.py [hunk 8]
```diff
@@ -490,15 +532,19 @@ def create_reference_genome(self,
...
-            metainfo.add_external_link(SEQUENCE_KEY, seq_link, text='Sequence data link')
+        self.__add_to_metainfo(metainfo, Metainfo.NAME, name, StringValue, required=True)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.ORGANISM, organism, StringValue)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.STRAIN, strain, StringValue)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.REFERENCE_GENOME_ASSEMBLY, assembly, StringValue)
+        self.__add_to_metainfo(metainfo, BioMetaKeys.REFERENCE_GENOME_RELEASE, release, StringValue)
+        self.__add_to_metainfo(metainfo, ANNOTATION_KEY, annotation_url, ExternalLink)
+        self.__add_to_metainfo(metainfo, metainfo.DESCRIPTION, description, StringValue)
+
+        if sequence_urls:
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: Why is it so complicated?  Old code was much clearer and more straightforward.

Comment (klas-genestack) [inline genestack_client/data_importer.py]: Why is it so complicated? Old code was much clearer and more straightforward.

File: genestack_client/data_importer.py (hunk 7 above)
Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: We just added clases of metainfo values, I did not think about this yet.

File: genestack_client/data_importer.py (hunk 8 above)
Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: `add_value` accepts only scalar value.
We want to pass list of values and make check for first item. 

It think about changing `add_value` to accept list as value, but not in this PR.

File: genestack_client/data_importer.py (hunk 7 above)
Comment (klas-genestack) [inline genestack_client/data_importer.py]: So, what's the purpose of this change (and similar ones)?  `add_external_link` calls `add_value` under the hood anyway, so I see no reason to make these changes.

File: genestack_client/data_importer.py (hunk 8 above)
Comment (klas-genestack) [inline genestack_client/data_importer.py]: I dislike this intensely.
1. Code looks ugly and way too complicated for the simple task.
2. I can't see why we test only the first URL.  And test so obliquely at that!  It's all urls in list, let's treat them all equally.

Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: > I can't see why we test only the first URL

Because if test fro first url is passed, we don't need to test other and cannot test other because this will provide an error.

I can move this test into  test function (if is_list ....) or wait a little and implement ListValue support

Comment (klas-genestack) [inline genestack_client/data_importer.py]: > we don't need to test other and cannot test other because this will provide an error

Why?

Comment (genestack-solomatin) [inline genestack_client/data_importer.py]: Add first with check

File: ChangeLog [hunk 9]
```diff
@@ -1,3 +1,6 @@
+* 1.13.0
```
Comment (klas-genestack) [inline ChangeLog]: 1.13.0?

**Comment by klas-genestack** [comment]
What's the status of this?

File: genestack_client/data_importer.py [hunk 10]
```diff
@@ -124,6 +125,40 @@ def __invoke_loader(self, parent, importer_type, metainfo):
         self.__process_links(metainfo)
         return self.importer.invoke('createFile', parent, importer_type, metainfo)['accession']
 
+    @staticmethod
+    def __add_to_metainfo(metainfo, key, value, value_type, required=False):
+        """
+        Add value to metainfo, check if value is not already present,
+        checks required fields.
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]:
```rst
Add ``key``: ``value`` pair to ``metainfo``.

Fails if ``metainfo`` already contains ``key`` or ``value`` is ``required``, but empty.
```

File: genestack_client/data_importer.py [hunk 11]
```diff
@@ -124,6 +125,40 @@ def __invoke_loader(self, parent, importer_type, metainfo):
...
+        """
+        current_value = metainfo.get(key)
+        if current_value is not None and value is not None:
+            raise GenestackException('Key "%s" is passed both as function argument '
+                                     'and inside metainfo object' % key)
+        if current_value:
+            return
+
+        if required and value is None:
+                raise GenestackException('Missing required key "%s", '
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: Bad indentation.

File: genestack_client/data_importer.py [hunk 12]
```diff
@@ -124,6 +125,40 @@ def __invoke_loader(self, parent, importer_type, metainfo):
...
+                                         'it should be passed as function argument '
+                                         'or in metainfo object' % key)
+        if value is not None:
```
Comment (klas-genestack) [inline genestack_client/data_importer.py]: But when `required == False` and `value is None` we don't add `key: None` to metainfo.
Is it expected behaviour?  Feels wrong to me, but if it is, it warrants a comment.

File: genestack_client/data_importer.py (hunk 10 above)
Comment (klas-genestack) [inline genestack_client/data_importer.py]: Add ``key``: ``value`` pair to ``metainfo``.

Fails if:
* ``metainfo`` already contains ``key``;
* ``key`` is ``required``, but no value is provided by user,
  and neither ``metainfo`` has already existing ``key`` (with some value).
